### DISCUSSION TITLE
Sending notifications via action node isn't working anymore

### DISCUSSION THREAD

**Discussion by neilenns** [zachowj/node-red-contrib-home-assistant-websocket#1591] [Q&A] [answered]
For years I've had an action node configured to send notifications to my Android device based on the incoming payload. It's configured like this:

![image](https://github.com/user-attachments/assets/6a1904d8-feab-4a22-b3cc-75d95d230bc1)

The incoming payload is created using a function node and specifies all the necessary info to send the notification:

```javascript
const newPayload = 
{
    "target": msg.payload.target,
    "data": {
        "title": msg.payload.title,
        "message": msg.payload.content,
        "data": {
            "push": {
                "thread-id": msg.payload.data.tag,
                "apns_headers":
                {
                    "apns-collapse-id": msg.payload.data.tag
                }
            },
            ...msg.payload.data
        }
    }
}
```

Recently I noticed I'm no longer receiving notifications, likely related to when I updated to a newer version of the websocket nodes. The error I'm getting is:

```
"ValidationError: "target" must be of type object"
```

It appears this is because of the `notify.{{payload.target}}`, which causes the node to show targets selection UI. If I replace it with `notify.mobile_app_my_phone` everything works.

What's the right way to specify a dynamic action now? I want to specify the phone to target as part of the incoming payload.

**Comment by zachowj** [comment, marked as answer]
The `msg.payload.target` property is an input for the action node and must follow a specific format. Since you're constructing the entire input before reaching the action node, you should include this in the action as well. For example:

```javascript
const newPayload = {
    action: `notify.${msg.payload.target}`,
    data: {
        title: msg.payload.title,
        message: msg.payload.content,
        data: {
            push: {
                "thread-id": msg.payload.data.tag,
                apns_headers: {
                    "apns-collapse-id": msg.payload.data.tag
                }
            },
            ...msg.payload.data
        }
    }
}
```

The other solution would be to ensure the `payload` object doesn’t include any supported input properties. You can review the list of allowed input properties [here](https://zachowj.github.io/node-red-contrib-home-assistant-websocket/node/action.html#input).

**Reply by neilenns** [reply]
Thanks for the quick response. That did the trick!
